### PR TITLE
Update the npm packages release process

### PR DESCRIPTION
**Pull request by youknowriad**
This PR updates the npm packages release process to match yesterday's discussion in the #core-js meeting.

See https://make.wordpress.org/core/2019/02/26/javascript-chat-summary-february-26th-2019/

### CODE REVIEW

File: docs/contributors/release.md
Comment (youknowriad) [inline docs/contributors/release.md]: I think creating this kind of patches for WordPress deserves its own section/doc.

File: docs/contributors/release.md
Comment (ntwb) [inline docs/contributors/release.md]: Currently the https://github.com/wordpress/wordpress-develop repo uses `x.y` for branches, and `x.y.z` for tags, this includes the major releases tagged as `x.y.z`, for example `5.1.0`

• https://github.com/WordPress/wordpress-develop/branches 
• https://github.com/WordPress/wordpress-develop/tags

I suggest rather than `wp/5.2` that `wp/5.2.0` is used so that Gutenberg uses a consistent 3 digit nomenclature?

File: docs/contributors/release.md
Comment (ntwb) [inline docs/contributors/release.md]: I see this as troublesome, not all packages get published during the _package release process_ and I'm not sure what the point of bumping the **minor** version in `major.minor.patch` achieves here.

Also, as we're hoping for wider spread usage from both inside and outside of the WordPress ecosystem bumping the **minor** version for no apparent reason is not very _semantic version_ friendly.

File: docs/contributors/release.md
Comment (ntwb) [inline docs/contributors/release.md]: FYI: To be aware of, WordPress currently backports when possible all the way back to WP 3.7

This raises the possibility that Gutenberg may also require patching back to `5.0`

For example, if a security issue is discovered later today and is proposed to be fixed in `5.1.1` it is conceivably possible that the _packages_ be updated will have to be backported to WordPress versions `5.0` and `5.1` as such the Gutenberg branches `wp/5.0`, `wp/5.1`, and `wp/trunk` will **all** require updating at this time.

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: I think Gutenberg should use the same thing `wp/5.2` for the branch and `x.y.z` for tags. the  `wp/5.2` will serve for all `5.2.1`, `5.2.2`...

File: docs/contributors/release.md
Comment (ntwb) [inline docs/contributors/release.md]: This will be problematic for other projects using our _packages_, it is not semver friendly at all.

See also my above comment https://github.com/WordPress/gutenberg/pull/14136/files#r260688733

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: The idea is that there will be potential bug fixes and security fixes we're not aware and we'll leave room for it. In general a Gutenberg release means enhancements... and not everything is mentioned in the CHANGELOG. 

With our lerna process, packages will be meant to be published without even touching them, if we change one of their dependencies.

Comment (youknowriad) [reply docs/contributors/release.md]: We're struggling because of this at the moment, we can't update trunk because there might be a 5.1.1 for WordPress requiring security fixes or bug fixes we're not aware of. Right now we default to patch when we don't know and it's blocking us.

I think this is the biggest improvement in this proposal

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: What alternatives do we have?

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: That's exactly why we need to leave "room" for bug fixes and security fixes in the package version number.

File: docs/contributors/release.md
Comment (ntwb) [reply docs/contributors/release.md]: When WordPress releases a major release it is both _branched_ and _tagged_, so a `5.0` branch is created, and a `5.1.0` tag is created:

• `5.0` branch https://github.com/WordPress/wordpress-develop/tree/5.1
• `5.1.0` tag  https://github.com/WordPress/wordpress-develop/tree/5.1.0

All `package.json` files use the 3 digit nomenclature:

https://github.com/WordPress/wordpress-develop/blob/5.1.0/package.json#L3:
> `"version": "5.1.0",`

Also, to note that the `5.0` branch has now had 4 minor releases so the _5.0 branch_ `package.json` file is now `5.0.4`:
https://github.com/WordPress/wordpress-develop/blob/5.0/package.json#L3:
> `"version": "5.1.0",`

----

If as you suggest, Gutenberg *only* use `wp/5.2` for all `5.2.x` releases I think that would work and be less work to maintain....

File: docs/contributors/release.md
Comment (ntwb) [reply docs/contributors/release.md]: Right, but shouldn't the team be made aware of pending security issues to accommodate this?

At this moment, if we were to publish an update to a _package_, and we bumped the _minor_ version there is no way of knowing that another security issue will be raised in the following days to which once again the _package_ will have to have the _minor_ bumped again and republished.

Will think on this some more overnight... 🤔

File: docs/contributors/release.md
Comment (ntwb) [reply docs/contributors/release.md]: Package version numbers cannot work like that as semantic versioning dictates that once `6.7.0` is published, `6.6.9` cannot be published, nothing less than `6.7.0` can be published.

File: docs/contributors/release.md
Comment (ntwb) [reply docs/contributors/release.md]: Continue to use semantic versioning:

If WordPress `5.1.1` is planned due to a security issue with the `@wordpress/editor` package which is at the time of writing is currently version `9.0.11` then the `@wordpress/editor` package is patched, updated, and published as version `9.1.0`.

That version `9.1.0` version of `@wordpress/editor` is then backported into both the `wp/5.0` and `wp/5.1` branches

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: Tags are automatically created by Lerna when we publish the packages. we can probably also tag `wp-5.2.1` ourselves, but the issue is that at the time of the npm release, we don't know yet if it's the final npm release that will be merged in this WP version and if we won't need more. So maybe just using the branch and the lerna tags is enough.

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: > semantic versioning dictates that once `6.7.0` is published, `6.6.9` cannot be published,

What makes you say that? How do npm package maintainers publish security fixes for old major and minor versions without forcing people to upgrade?

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: The thing is `9.1.0` might already exist because the `wp.5.2` might have already included it, so how do you go about patching the old versions without also adding features from the new versions?

File: docs/contributors/release.md
Comment (ntwb) [reply docs/contributors/release.md]: They don't, people are forced to upgrade to the version with the security fix.

For example, if a vulnerability is discovered in version `1.1.1` of a package, and that same package is now at version `7.8.9`, the fix for the `1.1.1` vulnerability must be released in either `7.8.10`, `7.9.0`, or `8.0.0`.

File: docs/contributors/release.md
Comment (ntwb) [reply docs/contributors/release.md]: For the past you cannot, the deprecations in the _packages_ will have to support the previous versions of WordPress per the Gutenberg deprecation policy

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: I think you're wrong, here's an example https://reactjs.org/blog/2018/08/01/react-v-16-4-2.html

Comment (nerrad) [reply docs/contributors/release.md]: This is where trying to align semantic versioning of packages with the WP release versioning clashes.  The problem here is that a latest release of a package could result in new features to the WP version the security fix is for.  This could be mitigated some by the use of the new feature flags functionality in GB but does complicate things.

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: I'm not talking about breaking any compatibility here. I'm taking about:

 - `WP 5.0.3` includes the editor package with `9.0.11`
 - `WP 5.1` includes the editor package with version `9.0.12`
 - `Trunk` includes the editor package with version `9.1.0`

(These are random numbers to illustrate)

A security issue is discovered and concerns all these packages, WordPress has to ship a security release for both `5.0`, `5.1` and also fix the issue in trunk. How do we proceed?

Contraints:

 - We want to include only this security fix for the previous versions, we don't want to add features that have been built for future versions.

Which versions do you publish and include in the WordPress releases `5.0.4`, `5.1.1` and trunk?

File: docs/contributors/release.md
Comment (nerrad) [reply docs/contributors/release.md]: > Package version numbers cannot work like that as semantic versioning dictates that once `6.7.0` is published, `6.6.9` cannot be published, nothing less than `6.7.0` can be published.

I just read through the [semver specs](https://semver.org/) and I'm not seeing this as something that is codified.  Is this something that is documented as a rule somewhere?

Comment (earnjam) [reply docs/contributors/release.md]: > Package version numbers cannot work like that as semantic versioning dictates that once `6.7.0` is published, `6.6.9` cannot be published, nothing less than `6.7.0` can be published.

I don't think the spec has an opinion on this, but I've never read it that way. If so, something like PHP isn't following semver. Maintaining LTS for a specific minor release wouldn't be possible. 

I always interpreted the spec as any individual version is just a comparison of the versions that are lower than it, regardless of release date.

File: docs/contributors/release.md
Comment (nosolosw) [inline docs/contributors/release.md]:
```suggestion
The Gutenberg repository mirrors the [WordPress SVN repository](https://make.wordpress.org/core/handbook/about/release-cycle/) in terms of branching for each SVN branch, a corresponding Gutenberg `wp/*` branch is created:
```

File: docs/contributors/release.md
Comment (ntwb) [reply docs/contributors/release.md]: @nerrad @earnjam I was not inferring a SemVer spec, I was inferring it as a restriction of the npm platform, that said:

@youknowriad I. Did. Not. Know. That. 💥

(I've just confirmed with npm that this can be done, also advised not to publish the _backported_ releases with the `latest` npm tag.

That will make things far simpler, though thinking about the "room" that we would leave behind, if version `1.1.1` of a package is already published, we bump the _minor_ so that it becomes version `1.2.0`, if we then have to patch version `1.1.1` for a vulnerability, by publishing the fix in `1.1.2` is that ok for semantic versioning? There is the possibility the "fix" breaks _backward compatibility which means the package should have been bumped by the _minor_ version and not the _patch_ version 🤔

File: docs/contributors/release.md
Comment (ntwb) [reply docs/contributors/release.md]: Based on the above https://github.com/WordPress/gutenberg/pull/14136#discussion_r260710630 knowing that we can backport the fixes going for the option of bumping the packages _minor_ version number before release would allow us to avoid the above scenario...

As the WP 5.0 branch would have v9.0.x of the package, before releasing WP 5.1 the package _minor_ is bumped to v9.1.x, this allows for multiple 9.0.x release to be backport puiblished for WP 5.0.x's, eg, `9.0.11` for WP `5.0.3`, `9.0.12` for WP `5.0.4`, `9.0.13` for WP `5.0.5 ` etc etc, and for WP 5.1.x, `9.1.1`, `9.1.2`, `9.1.3` and on are available, bump the _minor_ again for WP 5.2 and `9.2.x` range is available.

For projects other than WordPress using our packages, a note on all the repos that a _minor_ bump to all packages occurs when WP publishes a new version should document for the most part as to why we are slightly breaking semver on our packages, occasionly.

File: docs/contributors/release.md
Comment (ntwb) [reply docs/contributors/release.md]: Having that URL linked to "WordPress SVN repository" doesn't make much sense, using that link on the "in terms of branching for each SVN branch" text makes much more sense IMHO

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: > `9.0.12` for WP `5.0.4`

In my example, this is not possible because `9.0.12` already existed and is used in `5.1` and we don't have room to add fixes between `9.0.11` and `9.0.12`, that's why this is a special case using "metadata". I know that it's not perfect and that it's not "full semver" as npm will consider these versions as the same unless you explicitely target one but I don't think we have choices here because this scenario already happened, in this part, I'm trying to find a way to solve it when we face it.

Comment (ntwb) [reply docs/contributors/release.md]: Yes, I realise that, what I'm saying is we should do as you suggested above, and in the above scenario before WP 5.1 was shipped we would have bumped the editor package from the current `9.0.11` to `9.1.0`, thus WP 5.1 would have shipped with `9.1.0` and we'd have the remaining `9.0.x` versions for as many WP 5.0.x releases required over the coming years, WP 5.0.55 would use `9.0.66` 😏

**Comment by youknowriad** [comment]
Let's move forward here to unblock npm package releases, can I get a ✅?

**Review by gziolo — APPROVED** [review]
I had some minor notes only. In general, it seems to be solid and it aligns perfectly fine with what I have been thinking lately about this process. I was very disappointed that we couldn't publish the latest changes to npm since the very early January. All the changes proposed will make it all much more flexible and should allow us to do releases at will unless WordPress trunk is frozen. If that is still an issue we can iterate further.

I'm also not certain how Lerna and its version checks will work with what was proposed but this should be something we will learn ourselves pretty soon. I'm not concerned at all about major WordPress releases and sync from `master` branch as we need to enforce >= `minor` version bump for all packages. It should also fine for patch releases which will be performed from branches which will have Lerna commits included. The only issue I anticipate might happen is between Gutenberg releases if Lerna commit would get hash updated.

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: I prefer to keep it here, as it's important to read and stop the process if `trunk` is not open yet.

File: docs/contributors/release.md
Comment (youknowriad) [reply docs/contributors/release.md]: Maybe it can be reworded, but the reason it's here is to ensure that people are aware when to apply this flow before really applying it.

**Pull request merged**
